### ISSUE TITLE
MS Entra Validation

### ISSUE THREAD

**Issue by AxeOfMen** [sleelin/scimmy-routers#10]
I'm using the [SCIM validator](https://scimvalidator.microsoft.com/) provided by Microsoft for Entra (formerly Azure ActiveDirectory). 

I select "Discover My Schema" and set the SCIM endpoint and a test token value.
(It seems there is a bug with this validator. If you try to sign in using a work account you won't have the permissions needed and the recommendation is to sign in using a personal account)
I then see a page with "Schema discovered!" at the top and a number of properties listed. I click the "Test Schema" button and am presented this error:

"SCIM Validation failed. The attribute password for User is not supported by the SCIM protocol. Please refer to the SCIM RFC"

Here is my code
```js
const cors = require('cors');
const express = require('express');
const { default: SCIMMYRouters, SCIMMY } = require('scimmy-routers');

const app = express();
const PORT = 3030;

app.use(cors({origin: ['*']}));

app.use(function(req, res, next) {
  console.log(req.hostUrl,  req.url, req.rawHeaders);
  next();
});

SCIMMY.Resources.declare(SCIMMY.Resources.User)
  .ingress((resource, data) => {
    console.log(`Request to create/edit user ${JSON.stringify(resource)}, ${JSON.stringify(data)}`);
    let user = new SCIMMY.Schemas.User(data);
  })
  .egress((resource) => {
    console.log(`Request to read user ${JSON.stringify(resource)}`);
  })
  .degress((resource) => {
    console.log(`Request to delete user ${JSON.stringify(resource)}`);
  });

SCIMMY.Resources.declare(SCIMMY.Resources.Group)
  .ingress((resource, data) => {
    console.log(`Request to create/edit group ${JSON.stringify(resource)}, ${JSON.stringify(data)}`);
  })
  .egress((resource) => {
    console.log(`Request to read group ${JSON.stringify(resource)}`);
  })
  .degress((resource) => {
    console.log(`Request to delete group ${JSON.stringify(resource)}`);
  });

let authScheme = {
  type: "bearer",
  handler: (request) => {
    console.log("Inside auth check");
    if (!request.header("Authorization")?.startsWith("Bearer ")) {
      throw new Error("Authorization not detected!");
    } else {
      return 'WHATEVER';
    }
  }
};
app.use("/scim", new SCIMMYRouters(authScheme));

app.listen(PORT, (error) => {
    if (!error)
      console.log("Server is running on port " + PORT)
    else
      console.log("Server cannot start", error);
  }
);

```

Can I exclude the password from the schema?

**Comment by AxeOfMen** [comment]
Any pointers on fleshing my code out are welcome, too. It would lower the barrier to entry to have sample code showing ingress, egress and degress implementations.

**Comment by always-akshat** [comment]
Hey @AxeOfMen  did you figure this out? I am trying to fix a similar problem.

**Comment by sleelin** [comment]
Hi @AxeOfMen,
Apologies for the delay in getting back to you, I have been otherwise occupied.

It's possible to exclude the password attribute, or any other attribute, from the User schema using the schema definition's [truncate method](https://scimmyjs.github.io/SCIMMY.Types.SchemaDefinition.html#truncate) as below:
```js
SCIMMY.Schemas.User.definition.truncate("password");
```

Whilst this fixes the immediate error, unfortunately the MS Entra ID SCIM Validator is itself actually not compliant with the SCIM RFC. When I attempted validation after excluding the password attribute from the default schema, it then complained that the `emails[type eq "work"].display` attribute is invalid, which is entirely incorrect (see [RFC7643§8.7.1](https://datatracker.ietf.org/doc/html/rfc7643#section-8.7.1) for the actual specification). 

Ok, no big deal, right? We just need to remove the display attribute from emails, like we did for the password attribute.
```js
SCIMMY.Schemas.User.definition.truncate("password", "emails.display");
```

With that change, the validator then complains that the `emails[type eq "work"].primary` attribute is invalid, and to refer to the SCIM RFC again. This all starts to make sense when you consider that by default, MS Entra ID only actually uses a small subset of attributes from the default schemas. The trouble is, instead of ignoring the surplus attributes that aren't required, it complains that they aren't part of the SCIM RFC, which is confusing at least, and misleading at worst. Knowing this, if we remove the extra unsupported attributes, the validator no longer seems to take issue:
```js
SCIMMY.Schemas.User.definition.truncate([
    "password", "groups", "ims",
    "entitlements", "photos", "x509Certificates",
    "emails.display", "emails.primary",
    "phoneNumbers.display", "phoneNumbers.primary",
    "roles.display", "roles.primary"
]);
```

Great! Validation now passes, so we can move on to testing compliance. Here, we run into the final issue with the Entra ID SCIM Validation tool.

To test compliance, the tool sends out several requests to your SCIM endpoint, to make sure that the handling of GET/PUT/POST/PATCH/DELETE requests meets its requirements. Presently, using SCIMMY, _**all of the tests will fail**_, because as it turns out, all of those `primary` attributes actually _are_ part of the spec. Even though it _just_ claimed they were invalid, the tool is now including the `primary` attribute with every email, phone number, and role.

Because we excluded them from our schema to pass the first step in the schema validator, SCIMMY will now reject every one of the User creation compliance test requests with a `400 Bad Request` error, as they include the unknown attribute `primary`. This causes all of the subsequent User lookup/modification compliance test requests to also fail.

**TL;DR:** Microsoft's Entra ID SCIM Validation tool does not actually meet the SCIM RFC specification it so brazenly throws in your face. I would suggest raising the issue with Microsoft, but there doesn't seem to be any way to actually do this for the tool.

**Solution...?**
Alongside the tool, Microsoft does actually provide working and more comprehensive alternative means of testing your SCIM implementation via Postman, which is [detailed here](https://github.com/AzureAD/SCIMReferenceCode/wiki/Test-Your-SCIM-Endpoint#configure-postman). This Postman collection was actually something I used initially to help understand the protocol when starting this project, and I highly recommend it.

Thanks for your patience,
Sam

**Comment by brianpeiris** [comment]
@sleelin I understand your thinking here, in that you don't want to compromise SCIMMY's correct implementation to satisfy Microsoft's validation tools -- as you mentioned elsewhere, the Postman tests are also somewhat flawed. However, even if we put these validation tools aside, I believe Microsoft Entra itself has some of these non-compliance issues. So the reality is that people trying to use SCIMMY with Entra will run into them whether we like it or not.

At my previous employer, I had to [fork and modify SCIMMY](https://github.com/Cribcut/scimmy/pull/1) to work around Entra's flaws to actually get it to work in production, in addition to the truncations to you mention above. I wonder if it would be better for SCIMMY users if you could accommodate Entra somehow, even if they don't fix their implementation. My guess is that they may be open to fixing their validation tools, but it wouldn't mean much if Entra itself is still non-compliant -- and convincing them to fix that may not be possible, or at least take a long time on their end. It would be ideal if each SCIMMY user didn't have to maintain their own fork to support Entra.

So would you be open to accommodating Entra? I agree that it would be nice to keep the core of SCIMMY compliant, but one solution might be to make the core a bit more extensible or allow for overrides. That way you could keep your compliance by default, but add a layer on top of it to bend the rules for Entra's sake -- either by adding an `enableEntraCompatibility: true` flag that enables the override layer for Entra, or maintaining a separate `scimmy-entra-compatibility` shim extension/library that applies the overrides/extensibility for Entra.

Thoughts? I may be able to help implement such a thing if there's interest. Since I've already gone through the process of shipping Entra support in production, I'm familiar enough with it, and I'm able to run all the validation steps and manual testing in Entra as well.
